test/e2e/network/no_snat.go

```diff
@@ -52,70 +59,176 @@ var (
+		ginkgo.By("creating a test pod on one Node")
+		// use one node
+		node := nodes.Items[0]
+		// target Pod at Node
+
+		testPod := createTestPod(testPodPort2, false)
+		nodeSelection := e2epod.NodeSelection{Name: node.Name}
+		e2epod.SetNodeSelection(&testPod.Spec, nodeSelection)
+		f.PodClient().CreateSync(testPod)
+		ginkgo.By("creating a hostnetwork test pod on the same Node")
+		testPodHost := createTestPod(testPodPort2, true)
+		e2epod.SetNodeSelection(&testPodHost.Spec, nodeSelection)
+		f.PodClient().CreateSync(testPodHost)
 
 		ginkgo.By("sending traffic from each pod to the others and checking that SNAT does not occur")
 		pods, err := pc.List(context.TODO(), metav1.ListOptions{LabelSelector: noSNATTestName})
 		framework.ExpectNoError(err)
 
 		// hit the /clientip endpoint on every other Pods to check if source ip is preserved
-		// this test is O(n^2) but it doesn't matter because we only run this test on small clusters (~3 nodes)
 		for _, sourcePod := range pods.Items {
+			sourcePod := sourcePod
 			for _, targetPod := range pods.Items {
 				if targetPod.Name == sourcePod.Name {
 					continue
 				}
-				targetAddr := net.JoinHostPort(targetPod.Status.PodIP, testPodPort)
+				targetAddr := net.JoinHostPort(targetPod.Status.PodIP, testPodPort2)
+				ginkgo.By("testing from pod " + sourcePod.Name + " to pod " + targetPod.Name)
+
 				sourceIP, execPodIP := execSourceIPTest(sourcePod, targetAddr)
 				ginkgo.By("Verifying the preserved source ip")
 				framework.ExpectEqual(sourceIP, execPodIP)
 			}
 		}
+
+	})
+
+	ginkgo.It("Should be able to send traffic between Pods and HostNetwork pods without SNAT [LinuxOnly]", func() {
+		ginkgo.By("creating a test pod on each Node")
+		// create a hostNetwork pod in one of the nodes
+		testPodHost := createTestPod(testPodPort, true)
+		nodeSelection := e2epod.NodeSelection{Name: nodes.Items[0].Name}
+		e2epod.SetNodeSelection(&testPodHost.Spec, nodeSelection)
+		f.PodClient().CreateSync(testPodHost)
+
+		// create pods without hostNetwork in all nodes
+		testPod := createTestPod(testPodPort, false)
+		var wg sync.WaitGroup
+		for _, node := range nodes.Items {
+			// target Pod at Node
+			ginkgo.By("creating pod on node " + node.Name)
+			nodeSelection := e2epod.NodeSelection{Name: node.Name}
+			e2epod.SetNodeSelection(&testPod.Spec, nodeSelection)
+			wg.Add(1)
+			go func() {
+				defer ginkgo.GinkgoRecover()
+				defer wg.Done()
+				f.PodClient().CreateSync(testPod)
+			}()
+		}
+		wg.Wait()
+
+		ginkgo.By("sending traffic from each pod to the others and checking that SNAT does not occur")
+		pods, err := pc.List(context.TODO(), metav1.ListOptions{LabelSelector: noSNATTestName})
+		framework.ExpectNoError(err)
+
+		// hit the /clientip endpoint on every other Pods to check if source ip is preserved
+		for _, sourcePod := range pods.Items {
+			sourcePod := sourcePod
+			for _, targetPod := range pods.Items {
```

Comment:
You're testing in both directions here, but we I'm pretty sure we agreed that hostnetwork-to-pod is not required to use the pod IP as its source IP